src/engine/GrbMatrix.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Johannes Herrmann (johannes.roland.herrmann@mars.uni-freiburg.de)
+
+#pragma once
+
+extern "C" {
+#include <GraphBLAS.h>
+}
+
+#include <memory>
+#include <vector>
+
+// This class wraps the functionality of the GraphBLAS object GrB_Matrix.
+// Currently only boolean matrices are supported.
+class GrbMatrix {
+ private:
+  std::unique_ptr<GrB_Matrix> matrix_;
+  static bool isInitialized_;
+
+ public:
+  GrbMatrix(size_t numRows, size_t numCols) {
+    matrix_ = std::make_unique<GrB_Matrix>();
+    auto info = GrB_Matrix_new(matrix_.get(), GrB_BOOL, numRows, numCols);
+    handleError(info);
+  }
+
+  GrbMatrix() { matrix_ = std::make_unique<GrB_Matrix>(); }
+
+  // Move constructor
+  GrbMatrix(GrbMatrix&& otherMatrix) {
+    matrix_ = std::move(otherMatrix.matrix_);
+  };
+
+  // Disable copy constructor and assignment operator
+  GrbMatrix(const GrbMatrix&) = delete;
+  GrbMatrix& operator=(const GrbMatrix&) = delete;
+
+  ~GrbMatrix() { GrB_Matrix_free(matrix_.get()); }
+
+  GrbMatrix copy() const;
+
+  void setElement(size_t row, size_t col, bool value);
+
+  bool getElement(size_t row, size_t col) const;
+
+  // Create a matrix from the given lists of indices.
+  // For each given pair of indices, the corresponding entry in the result
+  // matrix is set to true. All other entries are false (by default).
+  static GrbMatrix build(const std::vector<size_t> rowIndices,
+                         const std::vector<size_t> colIndices, size_t numRows,
+                         size_t numCols);
+
+  // Create a square, diagonal matrix. All entries on the diagonal are set to
+  // true, all others to false. The resulting matrix will have nvals rows and
+  // columns.
+  static GrbMatrix diag(size_t nvals);
+
+  // Extract all true entries from the matrix. The first entry in the pair is
+  // the row index, the second entry is the column index.
+  std::vector<std::pair<size_t, size_t>> extractTuples() const;
+
+  // Extract a column from the matrix. Returns all row indices where this
+  // column's entries are true.
+  std::vector<size_t> extractColumn(size_t colIndex) const;
+
+  // Extract a row from the matrix. Returns all column indices where this
+  // rows's entries are true.
+  std::vector<size_t> extractRow(size_t rowIndex) const;
+
+  // Number of "true" values in the matrix.
+  size_t nvals() const;
+
+  size_t nrows() const;
+
+  size_t ncols() const;
```

Comment:
Call thos `numNonZero` `numRows` and `numCols` , just easier to parse for humans.